### PR TITLE
BTFS-1090: Added metadata add/update/delete operations

### PR DESCRIPTION
**Pull request by steveyeom**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

   1) Added metadata add/append/update/delete operation support on existing metadata 
      for a BTFS file in go-btfs.
   2) Added `btfs metadata add/rm` commands
   3) Added unit tests in core/coreunix level.

* **What is the current behavior?** (You can also link to an open issue here)
   No such functionalities in go-btfs.

* **What is the new behavior?** (You can also refer to a JIRA ticket here)
   BTFS-1090

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
   N/A since these are new features.

* **What dependencies / modules need to be updated as pre-requisites?** (Include relevant PR links here)
   https://github.com/TRON-US/go-unixfs/pull/16
   https://github.com/TRON-US/interface-go-btfs-core/pull/16
   https://github.com/TRON-US/go-btfs-chunker/commit/b60eabda2842f93f7645f839a72ddcff5b355a31
* **Description of changes**
   1. Added core/coreunix level new unit tests for each metadata operation: add/update/append/delete to
   `core/coreunix/test/metadata_test.go`
   2. Added/Modified modules in core/commands/ to support `btfs metadata add/rm` commands.
   `metadata.go`, `cat.go` and commands_test.go and root.go.
   3. Modified `unixfs.go` in core/coreapi/ to add new API methods for `btfs metadata add/rm` commands.
   4. Modified `metadata.go` and `add.go` to drive go-unixfs functions to support add/update/append/delete metadata operations.

---

* **Please check if the PR fulfills these requirements**
- [ x] The subject of this PR contains a JIRA ticket BTFS-xxxx (N/A for community PRs)
- [ x] PR description and commit messages are [good and meaningful](https://chris.beams.io/posts/git-commit/)
- [ x] Unit tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

* **Please make sure the following procedures have been applied before requesting reviewers**
- [x ] Code changes closely follow [Go Code Review Comments](https://github.com/golang/go/wiki/CodeReviewComments)
- [x ] Code has been *rebased* against latest master (`git merge` not recommended, unless you know what you are doing)
- [x ] Code changes have run through `go fmt`
- [ x] Code changes have run through `go mod tidy`
- [x ] All unit tests passed locally (`make test_go_test`)

### CODE REVIEW

File: core/coreunix/metadata.go [hunk 1]
```diff
@@ -1,20 +1,345 @@
...
+	mnode := children.MetaNode
+
+	// Retrieve SuperMeta info to build meta Dag modifier.
+	// Note that SuperMeta may or may not exist.
+	// TODO: unit test to check the cases..
+	b, err := GetMetaDataFromDagRoot(ctx, nd, modifier.dagService)
+	if err != nil {
+		return nil, err
+	}
+	var superMeta ihelper.SuperMeta
```
Comment (taiyangc) [inline core/coreunix/metadata.go]: why does removeMeta look at existing superMeta, but addMeta does not look at existing superMeta (if exists)

File: core/commands/metadata.go [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+		fileHash := req.Arguments[0]
+		tokenMetadata := req.Arguments[1]
+		// TODO: use for loop or batch for token metadata items.
+
+		p, err := api.Unixfs().AddMetadata(req.Context, path.New(fileHash), tokenMetadata, opts...)
+		if err != nil {
+			return err
+		}
+		h := ""
+		if p != nil {
```
[the remark this answers is not in the document]
Comment (steveyeom) [reply core/commands/metadata.go]: Fixed.

File: core/commands/metadata.go [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+		fileHash := req.Arguments[0]
+		tokenMetadata := req.Arguments[1]
+		// TODO: use for loop or batch for token metadata items.
+
+		p, err := api.Unixfs().RemoveMetadata(req.Context, path.New(fileHash), tokenMetadata, opts...)
+		if err != nil {
+			return err
+		}
+		h := ""
+		if p != nil {
```
[the remark this answers is not in the document]
Comment (steveyeom) [reply core/commands/metadata.go]: Fixed

File: core/coreapi/unixfs.go [hunk 4]
```diff
@@ -7,7 +7,9 @@ import (
 	"encoding/json"
 	"fmt"
 	ecies "github.com/TRON-US/go-eccrypto"
+	"github.com/TRON-US/go-unixfs/importer/helpers"
```
[the remark this answers is not in the document]
Comment (steveyeom) [reply core/coreapi/unixfs.go]: fixed

File: core/coreapi/unixfs.go [hunk 5]
```diff
@@ -258,6 +261,35 @@ func (api *UnixfsAPI) Get(ctx context.Context, p path.Path, metadata bool, opts
...
+
+			m := make(map[string]interface{})
+			err = json.Unmarshal(inb, &m)
+			if err != nil {
+				return nil, err
+			}
+
+			// Delete SuperMeta entries from the map.
+			var superMeta helpers.SuperMeta
+			typ := reflect.TypeOf(superMeta)
```
[the remark this answers is not in the document]
Comment (steveyeom) [reply core/coreapi/unixfs.go]: Talked with Eric. The reason for using reflect is not to have hard-coded key names here. 
If a new key is added to the `SuperMeta`, we need to add the new key manipulation code here if we use 
hard-coded key names.

File: core/coreunix/metadata.go [hunk 6]
```diff
@@ -1,20 +1,345 @@
...
+	if pi, ok := node.(*posinfo.FilestoreNode); ok {
+		node = pi.Node
+	}
+
+	mr, err := adder.mfsRoot()
+	if err != nil {
+		return nil, err
+	}
+	dir := gopath.Dir(path)
+	if dir != "." {
```
[the remark this answers is not in the document]
Comment (steveyeom) [reply core/coreunix/metadata.go]: Added comments.

File: core/coreunix/metadata.go [hunk 7]
```diff
@@ -1,20 +1,345 @@
...
+// Add the given `modifier.TokenMetadata` to the given `node`.
+func (modifier *MetaModifier) addMeta(nd ipld.Node) (ipld.Node, error) {
+	// Error cases
+	if nd == nil {
+		return nil, errors.New("Invalid argument: nil node value")
+	}
+
+	// Build Metadata DAG modifier for balanced format
+	// TODO: trickle format
+	ctx, closer := context.WithCancel(context.Background())
```
[the remark this answers is not in the document]
Comment (steveyeom) [reply core/coreunix/metadata.go]: fixed.

File: core/coreunix/metadata.go (hunk 1 above)
Comment (steveyeom) [reply core/coreunix/metadata.go]: Refined the code blocks.

**Comment by steveyeom** [comment]
Reflected all the review comments and all the unit tests passed.

File: core/coreunix/metadata.go [hunk 8]
```diff
@@ -275,69 +256,89 @@ func (modifier *MetaModifier) removeMeta(nd ipld.Node) (ipld.Node, error) {
...
 
-	// Add the given TokenMetadata to the metadata sub-DAG and return
-	// a new root of the BTFS file DAG.
-	newRoot, err := mdagmod.RemoveMetadata(nd, dr)
-	if err != nil {
-		return nil, err
+	return mdagmod, nil
+}
+
+func GetOrDefaultSuperMeta(b []byte, superMeta *ihelper.SuperMeta) error {
```
Comment (taiyangc) [inline core/coreunix/metadata.go]: Great

**Pull request merged**